HW2/P5/physics.pyx

```diff
@@ -69,17 +85,51 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+    # for j in range(i + 1, count):
+    #     XY2 = &(XY[j, 0])
+    #     V2 = &(V[j, 0])
+    #     if overlapping(XY1, XY2, R):
+    #         # SUBPROBLEM 4: Add locking
+    #         if not moving_apart(XY1, V1, XY2, V2):
+    #             collide(XY1, V1, XY2, V2)
+
+    #         # give a slight impulse to help separate them
+    #         for dim in range(2):
+    #             V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+    xgrid = <int> (XY[i, 0] / grid_spacing)
+    ygrid = <int> (XY[i, 1] / grid_spacing)
+    
+    # we only check x and y values greater than x in order to avoid colliding the same balls
+    # twice, once for lower values of x and y, then for the higher values
+    # check x, x+1 and x+2 
+    for x in range(xgrid, xgrid+3):
+        # check y values making sure that distance from original (x,y) is at most 2
+        for y in range(ygrid, ygrid+3-(x-xgrid)):
+            gridsize = <int> ((1.0 / grid_spacing) + 1)
+            # check that we are checking for valid grid squares, and that (x,y) is not the original
+            # square because then we might think that a ball is colliding with itself
+            # check that ballnumber is a valid ball
+            if x < gridsize and y < gridsize and (x != xgrid or y != ygrid) and Grid[x,y] < count:
+                j = Grid[x, y]
+                XY2 = &(XY[j, 0])
+                V2 = &(V[j, 0])
+                if overlapping(XY1, XY2, R):
+                    # SUBPROBLEM 4: Add locking
+                    acquire(&(locks[j]))
```

Comment:
Judging by the fact that this is your solution I think deadlock wasn't the issue, but it seems like locking and releasing in reverse order (i.e. lock i, j; release j, i) could cause such issues to happen. Maybe it is very unlikely (or impossible) in this problem setup to have 2 threads correspond to i,j and j,i indices.